### PR TITLE
RavenDB-7869 Enforce replication factor upon failure

### PR DESCRIPTION
**Pull request by karmeli87**

### CODE REVIEW

File: src/Raven.Server/ServerWide/Maintenance/ClusterObserver.cs [hunk 1]
```diff
@@ -252,18 +255,39 @@ class ClusterObserver : IDisposable
...
-                if (previousClusterStats.TryGetValue(mentorNode, out var mentorPrevClusterStats) == false ||
+
+                if (ShouldReassign(topology, promotable, dbName, current, previous))
+                {
+                    var node = FindLeastDbNode(promotable, dbName, current);
+                    if (node == null)
+                    {
+                        if (_logger.IsOperationsEnabled)
+                        {
+                            _logger.Operations($"Somthing is wrong with {dbName} on {promotable}, but we were unable to reassign it.");
```
Comment (ayende) [inline src/Raven.Server/ServerWide/Maintenance/ClusterObserver.cs:266]: This is nice and mysterious, and likely to repeat on every round, no?

File: src/Raven.Server/ServerWide/Maintenance/ClusterObserver.cs [hunk 2]
```diff
@@ -252,18 +255,39 @@ class ClusterObserver : IDisposable
...
+                        }
+                        continue;
+                    }
+                    if (_logger.IsOperationsEnabled)
+                    {
+                        _logger.Operations($"Somthing is wrong with {dbName} on {promotable}, so we reassign it to {node}.");
```
Comment (ayende) [inline src/Raven.Server/ServerWide/Maintenance/ClusterObserver.cs:272]: Mystery keep the ops teams on their toes. There is something wrong, but we wont tell you what.
Figure it out on your own.

File: src/Raven.Server/ServerWide/Maintenance/ClusterObserver.cs [hunk 3]
```diff
@@ -305,6 +329,56 @@ class ClusterObserver : IDisposable
...
+            var hasPrev = previous.TryGetValue(node, out var prevNodeStats) && prevNodeStats.LastReport.TryGetValue(db, out prevDbStats);
+
+            if (hasCurrent == false && hasPrev == false)
+                return true;
+
+            // Wait until we have more info
+            if (hasCurrent == false || hasPrev == false)
+                return false;
+
+            if (currentNodeStats.LastReportStatus != ClusterNodeStatusReport.ReportStatus.Ok &&
```
Comment (ayende) [inline src/Raven.Server/ServerWide/Maintenance/ClusterObserver.cs:353]: Need to take timing information into this. We don't immediately switch at the drop of a hat. We need to only switch when enough time (configurable) have passed.

File: src/Raven.Server/ServerWide/Maintenance/ClusterObserver.cs [hunk 4]
```diff
@@ -305,6 +329,56 @@ class ClusterObserver : IDisposable
             return false;
         }
 
+        private const DatabaseStatus BadFlags = Faulted | Shutdown;
```
Comment (ayende) [inline src/Raven.Server/ServerWide/Maintenance/ClusterObserver.cs:332]: Why is shutdown a bad flat?

File: src/Raven.Server/ServerWide/Maintenance/ClusterObserver.cs [hunk 5]
```diff
@@ -305,6 +329,56 @@ class ClusterObserver : IDisposable
...
+                prevNodeStats.LastReportStatus != ClusterNodeStatusReport.ReportStatus.Ok)
+                return true;
+
+            if ((currentDbStats.Status & BadFlags) != 0 && (prevDbStats.Status & BadFlags) != 0)
+                return true;
+
+            return false;
+        }
+
+        private string FindLeastDbNode(string badNode, string db,
```
Comment (ayende) [inline src/Raven.Server/ServerWide/Maintenance/ClusterObserver.cs:363]: Better method name, and need to explain why it is returning null.

File: src/Raven.Server/ServerWide/Maintenance/ClusterObserver.cs (hunk 2 above)
Comment (karmeli87) [reply src/Raven.Server/ServerWide/Maintenance/ClusterObserver.cs:272]: LoL!

**Pull request merged**